src/BlockDiagonals.jl

```diff
@@ -0,0 +1,338 @@
+    p = 0
+    while j > 0
+        p += 1
+        j -= ncols[p]
+    end
+    i -= sum(nrows[1:(p-1)])
+    # if row `i` outside of block `p`, set `p` to place-holder value `-1`
+    if i <= 0 || i > nrows[p]
+        p = -1
+    end
+    return p, i, j
+end
+
+Base.isapprox(M::AbstractMatrix, B::BlockDiagonal; kwargs...) = isapprox(B, M; kwargs...)
+function Base.isapprox(B::BlockDiagonal, M::AbstractMatrix; kwargs...)::Bool
+    return isapprox(Matrix(B), Matrix(M); kwargs...)
+end
+
+function Base.isapprox(B1::BlockDiagonal, B2::BlockDiagonal; kwargs...)
+    return isequal_blocksizes(B1, B2) && all(isapprox.(blocks(B1), blocks(B2); kwargs...))
+end
+
+function isequal_blocksizes(B1::BlockDiagonal, B2::BlockDiagonal)::Bool
+    return size(B1) == size(B2) && blocksizes(B1) == blocksizes(B2)
+end
+
+## Addition
+# TODO make type stable, maybe via Broadcasting?
+function Base.:+(B1::BlockDiagonal, B2::BlockDiagonal)
+    if size(B1) == size(B2) && size.(blocks(B1)) == size.(blocks(B2))
+        return BlockDiagonal(blocks(B1) .+ blocks(B2))
+    else
+        return Matrix(B1) + Matrix(B2)
+    end
+end
+
+Base.:+(M::AbstractMatrix, B::BlockDiagonal) = B + M
+Base.:+(B::BlockDiagonal, M::AbstractMatrix) = Matrix(B) + M
+
+function Base.:+(B::BlockDiagonal, M::StridedMatrix)::Matrix
+    size(B) == size(M) || throw(DimensionMismatch("dimensions must match"))
+    A = copy(M)
+    row = 1
+    for (j, block) in enumerate(blocks(B))
+        nrows = size(block, 1)
+        rows = row:(row + nrows-1)
+        A[rows, rows] .+= block
+        row += nrows
+    end
+    return A
+end
+
+Base.:+(M::Diagonal, B::BlockDiagonal) = B + M
+function Base.:+(B::BlockDiagonal, M::Diagonal)::BlockDiagonal
+    size(B) == size(M) || throw(DimensionMismatch("dimensions must match"))
+    A = copy(B)
+    d = diag(M)
+    row = 1
+    for (p, block) in enumerate(blocks(B))
+        nrows = size(block, 1)
+        rows = row:(row + nrows-1)
+        for (i, r) in enumerate(rows)
+            A[Block(p)][i, i] += d[r]
+        end
+        row += nrows
+    end
+    return A
+end
+
+Base.:+(M::UniformScaling, B::BlockDiagonal) = B + M
+function Base.:+(B::BlockDiagonal, M::UniformScaling)
+    return BlockDiagonal([block + M for block in blocks(B)])
+end
+
+## Multiplication
+Base.:*(n::Real, B::BlockDiagonal) = B * n
+Base.:*(B::BlockDiagonal, n::Real) = BlockDiagonal(n .* blocks(B))
+
+# TODO make type stable, maybe via Broadcasting?
+function Base.:*(B1::BlockDiagonal, B2::BlockDiagonal)
+    if size(B1) == size(B2) && size.(blocks(B1)) == size.(blocks(B2))
+        return BlockDiagonal(blocks(B1) .* blocks(B2))
+    else
+        return Matrix(B1) * Matrix(B2)
+    end
+end
+
+function Base.:*(B::BlockDiagonal, M::AbstractMatrix)::Matrix
+    if size(B, 2) !== size(M, 1)
+        throw(DimensionMismatch("A has dimensions $(size(B)) but B has dimensions $(size(M))"))
+    end
+    st = 1  # start
+    ed = 0  # end
+    d = []
```

Comment:
give this a type.
Or rephrase the below into
`d = map(blocks) do block`

Reply:
changed -- i still don't understand `do` syntax properly so i just did `d = Matirx[]`

Reply:
`do` blocks are just a special way to tell the function call that its first argument is a function. All of the following are equivalent:
```julia
map(x) do xi
    ...
end

map(xi -> begin
    ...
end, x)

map(function (xi)
    ...
end, x)
```